indexing-service/src/main/java/org/apache/druid/indexing/common/task/AbstractBatchIndexTask.java

```diff
@@ -398,22 +375,21 @@ private boolean tryLockWithDetermineResult(TaskActionClient client, LockGranular
 
   protected boolean tryTimeChunkLock(TaskActionClient client, List<Interval> intervals) throws IOException
   {
-    // The given intervals are first converted to align with segment granularity. This is because,
-    // when an overwriting task finds a version for a given input row, it expects the interval
-    // associated to each version to be equal or larger than the time bucket where the input row falls in.
-    // See ParallelIndexSupervisorTask.findVersion().
-    final Set<Interval> uniqueIntervals = new HashSet<>();
+
+    // In this case, the intervals to lock must be aligned with segmentGranularity if it's defined
+    final Iterator<Interval> intervalIterator;
     final Granularity segmentGranularity = getSegmentGranularity();
-    for (Interval interval : intervals) {
-      if (segmentGranularity == null) {
-        uniqueIntervals.add(interval);
-      } else {
-        Iterables.addAll(uniqueIntervals, segmentGranularity.getIterable(interval));
-      }
+    if (segmentGranularity == null) {
+      Set<Interval> uniqueIntervals = new HashSet<>(intervals);
+      ArrayList<Interval> condensedIntervals = JodaUtils.condenseIntervals(() -> uniqueIntervals.iterator());
+      intervalIterator = condensedIntervals.iterator();
+    } else {
+      IntervalsByGranularity intervalsByGranularity = new IntervalsByGranularity(intervals, segmentGranularity);
```

Comment:
Are we no longer condensing the intervals if segmentGranularity != null?

Reply:
I will fix this in the next commit